src/provider/provider_os_memory.c

```diff
@@ -0,0 +1,299 @@
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <assert.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#include <umf.h>
+#include <umf/memory_provider_ops.h>
+#include <umf/providers/provider_os_memory.h>
+
+typedef struct os_memory_provider_s {
+    umf_os_memory_provider_params_t params;
+} os_memory_provider_t;
+
+static __thread int32_t TLS_last_native_error;
+
+static const char *Native_error_str[] = {
+    "Success",                            // UMF_OS_RESULT_SUCCESS
+    "wrong alignment (not a power of 2)", // UMF_OS_RESULT_ERROR_WRONG_ALIGNMENT
+    "NUMA binding mode not supported for VisibilityShared", // UMF_OS_RESULT_ERROR_NUMA_MODE_NOT_SUPPORTED
+    "mmap() failed",                       // UMF_OS_RESULT_ERROR_MMAP_FAILED
+    "mmap() returned address not aligned", // UMF_OS_RESULT_ERROR_MMAP_ADDR_NOT_ALIGNED
+    "mbind() failed",                      // UMF_OS_RESULT_ERROR_MBIND_FAILED
+    "munmap() failed",                     // UMF_OS_RESULT_ERROR_MUNMAP_FAILED
+    "madvise(MADV_FREE) failed",     // UMF_OS_RESULT_ERROR_MADVISE_FREE_FAILED
+    "madvise(MADV_DONTNEED) failed", // UMF_OS_RESULT_ERROR_MADVISE_DONTNEED_FAILED
+};
+
+static enum umf_result_t copy_nodemask(const unsigned long *nodemask,
+                                       unsigned long maxnode,
+                                       unsigned long **outNodemask) {
+    if (maxnode == 0 || nodemask == NULL) {
+        *outNodemask = NULL;
+        return UMF_RESULT_SUCCESS;
+    }
+
+    size_t nodemask_bytes = (maxnode / 8) + ((maxnode % 8) ? 1 : 0);
+    size_t ul_size = sizeof(unsigned long);
+
+    // round to the next multiple of sizeof(unsigned long)
+    if (nodemask_bytes % ul_size != 0 || nodemask_bytes == 0) {
+        nodemask_bytes += ul_size - nodemask_bytes % ul_size;
+    }
+
+    *outNodemask = (unsigned long *)calloc(1, nodemask_bytes);
+    if (!*outNodemask) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    memcpy(*outNodemask, nodemask, nodemask_bytes);
+
+    return UMF_RESULT_SUCCESS;
+}
+
+enum umf_result_t os_initialize(void *params, void **provider) {
+    if (provider == NULL || params == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    os_memory_provider_t *os_provider =
+        (os_memory_provider_t *)calloc(1, sizeof(os_memory_provider_t));
+    if (!os_provider) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    umf_os_memory_provider_params_t *os_params =
+        (umf_os_memory_provider_params_t *)params;
+    os_provider->params = *os_params;
+
+    enum umf_result_t ret = copy_nodemask(
+        os_params->nodemask, os_params->maxnode, &os_provider->params.nodemask);
+    if (ret != UMF_RESULT_SUCCESS) {
+        return ret;
+    }
+
+    *provider = os_provider;
```

Comment:
How should I check the params? What do you mean by that?

Reply:
We should check if all params passed by user are correct e.g. int protection is in the valid range etc. We could even call a map + bind some small allocation here and then do a cleanup, but I'm not sure what others think about it?

Reply:
I'm not sure about mmap + bind in initialize() because even if it succeeds now, it can fail later (and vice versa). We definitely can add some checks for parameters here and return an early error if we detect unsupported combination.

Reply:
Yes I think we should fail during creation of the provider once instead of failing each time  user try to call alloc() and use the provider that contains bad parameters (that are immutable).